specification/evaluation-context/evaluation-context.md

```diff
@@ -1,5 +1,42 @@
 # Evaluation Context
 
+**Status**: [Experimental](../README.md#document-statuses)
+
 ## Overview
 
-// TODO: this is a stub.
+The `evaluation context` provides ambient information for the purposes of flag evaluation. Contextual data may be used as the basis for targeting, including rule-based evaluation, overrides for specific subjects, or fractional flag evaluation.
+
+### Fields
+
+##### Requirement 3.1
+
+> The `evaluation context` structure **MUST** define a required `targeting key` field of type string, identifying the subject of the flag evaluation.
+
+The targeting key uniquely identifies the subject (end-user, or client service) of a flag evaluation. Providers may use this as a unique identifier for fractional-flags, rules, or overrides targeting specific users.
+
+##### Requirement 3.2
+
+> The evaluation context **MUST** define the following optional fields: `email` (string), `first name` (string), `last name`(string), `name`(string), `ip`(string), `tz`(string), `locale`(string), `country code` (string), `timestamp`(date), `environment`(string), `application`(string), and `version`(string).
```

Comment:
Can you say a little more about the interplay between MUST and optional?  Like.. it has to return null if something were to ask for that value?

Reply:
> Can you say a little more about the interplay between MUST and optional? Like.. it has to return null if something were to ask for that value?

It has to be defined, but can be empty/null. In Java, I would expect some getters/setters for these values, I think, as well as some other place to store additional attributes (some HashMap, perhaps).

Reply:
Does this mean that the client can do something like context.getAttribute("ip") and expect null if the provider does not support this? Or does the context need a specific API for for each of these attributes e.g. context.getEmailAttribute()?

Reply:
> Does this mean that the client can do something like context.getAttribute("ip") and expect null if the provider does not support this? Or does the context need a specific API for for each of these attributes e.g. context.getEmailAttribute()?

@s-sen  Continuing with a hypothetical Java implementation, I think for improved discoverability, it would be ideal if the context had a `context.getEmailAttribute()`. This way, the compiler/editor can help developers explore the standard fields, and even potentially give them creative ideas for rule-based flags. It also implicitly shows that these attributes are "first class". All additional custom attributes would work like `context.getCustomAttribute("myCustomAttr")`. That's my thinking.

@justinabrahms

Reply:
@toddbaert I suppose I'm not entirely sold on having these as 'first class' attributes. I'm concerned about why we need them to be first class. 

I don't want developers to see this as a TODO list (especially b/c I'd prefer our developers to not be shoving in PII like the names of users into the flagging system). 

If providers need this information, I'm concerned that it will bloat over time and be really noisy. If they don't need this information.. then I'm not sure why we're dictating these fields.

Reply:
@justinabrahms I think it goes back to the fact that some providers DO have these as first class fields (launch darkly). Since they do, I think we need it in OpenFeature as well. If we don't, we end up in the following hypothetical situation:

- Launch Darkly defines `ip` for the location of the `ip` field in their SDK.
- In their `provioder` doc, they say authors should populate a field in the OpenFeaure provider called `ip`, and their provider will pull this value from there.
- Application Author does as instructed above
- Application Author switches to a new custom provider, `provider x`, which dictates the ip should be defined in a field called `remote_addr`.
- Application Author must update all their code to now put the IP here.

If this was a "well-defined" field, each provider would "know" where to get these values, and we have no concerns switching between providers.

Reply:
I better understand what you mean. These keys are for providers to agree upon so they can map their own sdk attributes to them. I'm fine with this for now, but would like to talk about PII in the future. 

https://github.com/open-feature/spec/issues/64

Reply:
@toddbaert from the LaunchDarkly perspective these are optional fields and agree with @justinabrahms on not having them as `first class` attributes, we leave them up to the customer to decide on usage and marking private.